Installation fails with npm v7

~~For some reason, `npm install -g npq` fails during the postinstall script when using Node 14.16.0. In my case Node is installed with nvm, however, the npq install works fine on 14.15.4.~~

Edit (see [update below](https://github.com/lirantal/npq/issues/185#issuecomment-815033188)): `npm install -g npq` fails during the postinstall script when using npm v7.

## Expected Behavior
Installation leads to the usual "Thank you for installing npq... do you want to add aliases?" prompt. This works fine with Node 14.16.1 and npm 6.14.12.

## Current Behavior

I've tried both a local install (without -g) and the normal global install, with the same result. The prompt text is printed in the middle of the error, before it exits:

```
npm ERR! path /Users/andron/.nvm/versions/node/v14.16.0/lib/node_modules/npq
npm ERR! command failed
npm ERR! signal SIGINT
npm ERR! command sh -c node scripts/postinstall.js
npm ERR! Thank you for installing npq! We want to help you make conscious decisions before installing potentially dangerous packages.
npm ERR! To do that, we can alias npm and yarn to npq, so that e.g. `npm install <package>` will first use npq to verify the package and prompt you if it finds any issues.
npm ERR! ? Do you want to add zsh aliases for npm and yarn? (Y/n) 

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/andron/.npm/_logs/2021-04-07T14_08_12_889Z-debug.log
```

Relevant lines from the debug log:

```
1016 info run npq@2.0.16 postinstall node_modules/npq node scripts/postinstall.js
1017 info run npq@2.0.16 postinstall { code: null, signal: 'SIGINT' }
1018 timing reify:rollback:createSparse Completed in 670ms
1019 timing reify:rollback:retireShallow Completed in 0ms
1020 timing command:install Completed in 5030ms
1021 verbose stack Error: command failed
1021 verbose stack     at ChildProcess.<anonymous> (/Users/andron/.nvm/versions/node/v14.16.0/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/index.js:64:27)
1021 verbose stack     at ChildProcess.emit (events.js:315:20)
1021 verbose stack     at maybeClose (internal/child_process.js:1048:16)
1021 verbose stack     at Socket.<anonymous> (internal/child_process.js:439:11)
1021 verbose stack     at Socket.emit (events.js:315:20)
1021 verbose stack     at Pipe.<anonymous> (net.js:673:12)
1022 verbose pkgid npq@2.0.16
1023 verbose cwd /Users/andron
1024 verbose Darwin 19.6.0
1025 verbose argv "/Users/andron/.nvm/versions/node/v14.16.0/bin/node" "/Users/andron/.nvm/versions/node/v14.16.0/bin/npm" "install" "-g" "npq"
1026 verbose node v14.16.0
1027 verbose npm  v7.8.0
1028 error path /Users/andron/.nvm/versions/node/v14.16.0/lib/node_modules/npq
1029 error command failed
1030 error signal SIGINT
1031 error command sh -c node scripts/postinstall.js
1032 error Thank you for installing npq! We want to help you make conscious decisions before installing potentially dangerous packages.
1032 error To do that, we can alias npm and yarn to npq, so that e.g. `npm install <package>` will first use npq to verify the package and prompt you if it finds any issues.
1032 error ? Do you want to add zsh aliases for npm and yarn? (Y/n)
1033 verbose exit 1
```

## Your Environment **(edited per update)**:
* Version used: 2.0.16
* Environment name and version:
    * Tested with Node.js v14.16.0, 14.16.1, and 15.14.0,
    * Bug occurs with npm 7.7.6 and 7.8.0 (haven't tested further)
    * installed with nvm 0.38.0,
    * zsh shell
* Operating System and version (desktop or mobile): macOS catalina 10.15.7

Installed with silly log level, getting these:

```
npm info run npq@2.0.19 postinstall node_modules/npq node scripts/postinstall.js
npm info run npq@2.0.19 postinstall { code: null, signal: 'SIGINT' }
npm timing reify:rollback:createSparse Completed in 611ms
npm timing reify:rollback:retireShallow Completed in 0ms
npm timing command:install Completed in 6032ms
npm verb stack Error: command failed
npm verb stack     at ChildProcess.<anonymous> (/Users/lirantal/.nvm/versions/node/v15.8.0/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/index.js:64:27)
npm verb stack     at ChildProcess.emit (node:events:378:20)
npm verb stack     at maybeClose (node:internal/child_process:1067:16)
npm verb stack     at Socket.<anonymous> (node:internal/child_process:453:11)
npm verb stack     at Socket.emit (node:events:378:20)
npm verb stack     at Pipe.<anonymous> (node:net:665:12)
npm verb pkgid npq@2.0.19
npm verb cwd /private/tmp/a
npm verb Darwin 19.6.0
npm verb argv "/Users/lirantal/.nvm/versions/node/v15.8.0/bin/node" "/Users/lirantal/.nvm/versions/node/v15.8.0/bin/npm" "install" "npq" "--loglevel" "silly"
npm verb node v15.8.0
npm verb npm  v7.5.1
npm ERR! path /private/tmp/a/node_modules/npq
npm ERR! command failed
npm ERR! signal SIGINT
npm ERR! command sh -c node scripts/postinstall.js
npm ERR! Thank you for installing npq! We want to help you make conscious decisions before installing potentially dangerous packages.
npm ERR! To do that, we can alias npm and yarn to npq, so that e.g. `npm install <package>` will first use npq to verify the package and prompt you if it finds any issues.
npm ERR! ? Do you want to add zsh aliases for npm and yarn? (Y/n)
npm verb exit 1
npm timing npm Completed in 6245ms
```

--- a/__tests__/scripts.test.js
+++ b/__tests__/scripts.test.js
@@ -13,11 +13,14 @@ inquirer.prompt = jest.fn().mockResolvedValue({ install: true })
 
 let mockGetShellConfig
 let mockIsRunningInYarn
+let mockGetNpmVersion
 
 beforeEach(() => {
   mockGetShellConfig = jest.spyOn(helpers, 'getShellConfig').mockImplementation(() => ({ name: 'testShell', profilePath: TEST_PROFILE_PATH, aliases: TEST_ALIAS }))
   // Our install script does not run when installing with yarn, but we still want to be able to run tests with yarn
   mockIsRunningInYarn = jest.spyOn(helpers, 'isRunningInYarn').mockImplementation(() => false)
+  // Our install script does not run when installing with npm V7, but we still want to be able to run tests with all npm versions
+  mockGetNpmVersion = jest.spyOn(helpers, 'getNpmVersion').mockImplementation(() => '6.0.0')
 })
 
 afterEach(async () => {
@@ -54,6 +57,22 @@ test('postinstall script does not run in yarn', async () => {
   expect(spy).not.toHaveBeenCalled()
 
   process.env.npm_execpath = oldExecPath
+  spy.mockRestore()
+})
+
+test('postinstall script does not run in npm v7', async () => {
+  // npm v7 does not allow for stdin input during `npm install [-g] npq`
+  mockGetNpmVersion.mockRestore()
+  const oldUserAgent = process.env.npm_config_user_agent
+
+  const spy = jest.spyOn(console, 'log')
+  // pretend that we're running in npm v7, whether or not we actually are
+  process.env.npm_config_user_agent = 'npm/7.19.1 node/v14.16.1 win32 x64 workspaces/false'
+
+  await postinstall.runPostInstall()
+  expect(spy).not.toHaveBeenCalled()
+
+  process.env.npm_config_user_agent = oldUserAgent
   spy.mockRestore()
 })
 

--- a/scripts/postinstall.js
+++ b/scripts/postinstall.js
@@ -11,7 +11,7 @@ const runPostInstall = async () => {
     return
   }
 
-  if (semver.gte(helpers.getNpmVersion(), '7.x.x')) {
+  if (semver.gte(helpers.getNpmVersion(), '7.0.0')) {
     // `npm install` in npm v7 cannot get stdin input, so we can't run this script there
     return
   }

--- a/scripts/postinstall.js
+++ b/scripts/postinstall.js
@@ -1,12 +1,18 @@
 const fs = require('fs')
 const chalk = require('chalk')
 const inquirer = require('inquirer')
+const semver = require('semver')
 
 const helpers = require('./scriptHelpers')
 
 const runPostInstall = async () => {
   if (helpers.isRunningInYarn()) {
     // `yarn add` cannot get stdin input, so we can't run this script there
+    return
+  }
+
+  if (semver.gte(helpers.getNpmVersion(), '7.x.x')) {
+    // `npm install` in npm v7 cannot get stdin input, so we can't run this script there
     return
   }
 

--- a/scripts/scriptHelpers.js
+++ b/scripts/scriptHelpers.js
@@ -47,6 +47,11 @@ module.exports.isRunningInYarn = () => {
   return binaryName.toLowerCase().includes('yarn')
 }
 
+module.exports.getNpmVersion = () => {
+  const npmData = process.env['npm_config_user_agent'] || ''
+  return npmData.split(' ')[0].split('/')[1]
+}
+
 const getProfile = async (profilePath) => {
   try {
     // eslint-disable-next-line security/detect-non-literal-fs-filename

--- a/scripts/scriptHelpers.js
+++ b/scripts/scriptHelpers.js
@@ -1,6 +1,7 @@
 const fs = require('fs')
 const os = require('os')
 const path = require('path')
+const semver = require('semver')
 
 const BASH_ZSH_ALIASES = '\nalias npm="npq-hero"\nalias yarn="NPQ_PKG_MGR=yarn npq-hero"\n'
 const SHELLS = {
@@ -48,8 +49,9 @@ module.exports.isRunningInYarn = () => {
 }
 
 module.exports.getNpmVersion = () => {
-  const npmData = process.env['npm_config_user_agent'] || ''
-  return npmData.split(' ')[0].split('/')[1]
+  const npmData = process.env['npm_config_user_agent'] || '0.0.0'
+  const version = /npm\/(.*) node/.exec(npmData)[1]
+  return semver.valid(version) ? version : '0.0.0'
 }
 
 const getProfile = async (profilePath) => {

